src/main/java/org/jenkinsci/plugins/workflow/cps/global/CompilationErrorException.java

```diff
@@ -0,0 +1,18 @@
+package org.jenkinsci.plugins.workflow.cps.global;
+
+import org.codehaus.groovy.control.MultipleCompilationErrorsException;
+
+import java.io.NotSerializableException;
+
+/**
+ * An exception that replaces Groovy's {@link MultipleCompilationErrorsException},
+ * because that is not serializable (which would, under certain circumstances,
+ * lead to the user being presented with a {@link NotSerializableException} instead
+ * of a compilation error -- see JENKINS-40109).
+ */
+public class CompilationErrorException extends RuntimeException {
+    public CompilationErrorException(MultipleCompilationErrorsException original) {
+        super(original.toString());
+        setStackTrace(original.getStackTrace());
```

Comment:
The call to `super()` will call `fillInStackTrace` to create a stack trace for the new exception, and since you are setting the stack trace explicitly here I think you should override `fillInStackTrace` so that it just returns `this` and doesn't do any actual work.